### PR TITLE
[channel-client] Fake channel client unit test

### PR DESCRIPTION
**Pull request by cf19drofxots**
PR addressing second item in issue https://github.com/statechannels/monorepo/issues/688

(Re-opened the previously closed [PR](https://github.com/statechannels/monorepo/pull/750) as work is still in draft)

### CODE REVIEW

File: packages/channel-client/src/fake-channel-client.ts [hunk 1]
```diff
@@ -12,21 +12,19 @@ import {
   Allocation,
   PushMessageResult
 } from './types';
-import {sleep} from './utils';
 
-const FAKE_DELAY = 100; // ms
+export const CHANNEL_ID = '0x1234';
```
Comment (snario) [inline packages/channel-client/src/fake-channel-client.ts]: Doesn't this mean we can only create one channel with this fake client? This should probably be like the hash of the arguments or something like that.

File: packages/channel-client/tsconfig.json [hunk 2]
```diff
@@ -4,7 +4,8 @@
     "baseUrl": "src",
     "module": "commonjs",
     "outDir": "lib",
-    "target": "es2015"
+    "target": "es2015",
+    "esModuleInterop": true
```
Comment (snario) [inline packages/channel-client/tsconfig.json]: If you can get away without adding this, that would be nice.

**Comment by cf19drofxots** [comment]
Status update: I've been trying to figure out the dynamic of messaging passing between the clients in order to write more appropriate tests for the `FakeChannelClient` but after not making much progress @tomclose and I synced on the expected behavior from the `ChannelClientInterface`.

cc @snario

File: packages/rps/src/redux/auto-opponent/__tests__/auto-opponent.test.ts [hunk 3]
```diff
@@ -17,7 +17,7 @@ const reducer = combineReducers({
 });
 
 // TODO: Get this working with a mock for channel-client
-it(
+it.skip(
```
Comment (cf19drofxots) [inline packages/rps/src/redux/auto-opponent/__tests__/auto-opponent.test.ts]: Skipping to get CI green. This has to be unskipped and is on the radar for how this is supposed to behave with the fake channel client.

File: packages/channel-client/src/fake-channel-client.ts [hunk 4]
```diff
@@ -1,6 +1,6 @@
-import {Wallet} from 'ethers';
 import {bigNumberify} from 'ethers/utils';
 import EventEmitter = require('eventemitter3');
+import log = require('loglevel');
```
Comment (snario) [inline packages/channel-client/src/fake-channel-client.ts]: Why use import assignment here, is it necessary?

File: packages/channel-client/src/fake-channel-client.ts [hunk 5]
```diff
@@ -56,33 +58,37 @@ export class FakeChannelClient implements ChannelClientInterface<ChannelResult>
...
-    const latestState = this.findChannel(channelId);
+    this.updatePlayerIndex(1);
+    log.debug(`Player ${this.playerIndex} joining channel ${channelId}`);
+    await this.verifyTurnNum(this.latestState!.turnNum);
+
     // skip funding by setting the channel to 'running' the moment it is joined
     // [assuming we're working with 2-participant channels for the time being]
-    this.latestState = {...latestState, turnNum: bigNumberify(3).toString(), status: 'running'};
+    this.latestState = {
+      ...this.latestState!,
```
Comment (snario) [inline packages/channel-client/src/fake-channel-client.ts]: Why do you need to use `!` here? Can you avoid using it?

File: packages/channel-client/src/fake-channel-client.ts [hunk 6]
```diff
@@ -56,33 +58,37 @@ export class FakeChannelClient implements ChannelClientInterface<ChannelResult>
...
+      turnNum: bigNumberify(3).toString(),
```
Comment (snario) [inline packages/channel-client/src/fake-channel-client.ts]: Does `toString` do the same thing as `toHexString`?

File: packages/channel-client/src/utils.ts [hunk 7]
```diff
@@ -1,3 +1,25 @@
...
+
+export function calculateChannelId(
+  participants: Participant[],
+  allocations: Allocation[],
+  appDefinition: string,
+  appData: string
+): string {
+  // This `message` construction revolves around `hashMessage` taking in a
+  // string but since the concat of the arguments are unique to each channel
+  // the channel ID should be different as a result.
```
Comment (snario) [inline packages/channel-client/src/utils.ts]: Why should it be different per `allocations` or `appData`? Channel ID is unique on `participants`, `channelNonce`, and `appDefinition` only, and there is a specific calculation for it. I think there are helper utils in `devtools`.

File: packages/channel-client/tests/constants.ts [hunk 8]
```diff
@@ -0,0 +1,10 @@
+import {AddressZero, Zero, One} from 'ethers/constants';
+
+// These could be generated on the fly but why run unecessary compute cycles?
```
Comment (snario) [inline packages/channel-client/tests/constants.ts]: Unnecessary comment

File: packages/channel-client/tests/fake-channel-client.test.ts [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+    });
+  });
+
+  it('instantiates', () => {
+    expect(clientA).toBeDefined();
+    expect(clientB).toBeDefined();
+  });
+
+  describe('creates a channel', () => {
+    expect.assertions(2);
```
Comment (snario) [inline packages/channel-client/tests/fake-channel-client.test.ts]: Why is this necessary?

File: packages/channel-client/tests/fake-channel-client.test.ts [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+
+    clientB.onChannelProposed((result: ChannelResult) => {
+      clientBEventEmitter.emit('ChannelProposed', result);
+    });
+  });
+
+  it('instantiates', () => {
+    expect(clientA).toBeDefined();
+    expect(clientB).toBeDefined();
+  });
```
Comment (snario) [inline packages/channel-client/tests/fake-channel-client.test.ts]: I think this test is redundant because `beforeEach` would explode if neither were defined.
